KEP-0661: StatefulSet volume resize (#2842)
kubernetes/enhancements: keps/sig-storage/0661-statefulset-volume-resize/README.md

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

Currently kubernetes supports volume expansion as a beta feature since v1.16. However, 
this feature support is only limited to standalone pods and deployments. Expansion of 
volumes associated with a Statefulset is not supported. This enhancement proposes to add 

> the ability to resize volumes associated with a StatefulSet via modifications to it's specification.

Reviewer: it's  -> its

Users will be able to recover from volume expansion failures using the same mechanism 

> as listed in [KEP 1790](../1790-recover-resize-failure/README.md).

Reviewer: Can you refer to this updated KEP?  https://github.com/kubernetes/enhancements/pull/2807

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### API server validation relaxation
Modifications would be made to `ValidateStatefulSetUpdate` inorder to allow changes in the
storage size of the volume in the `VolumeClaimTemplates`. The new requested size will be validated to 
prevent a user from shrinking the volume.

### StatefulSet controller changes
The `sync` function in the `StatefulSetController` runs the reconciliation to match the current
reality and the expected declaration in the `Spec`. An addendum will be made to validate each
PVC associated with each pod and ensure that their sizes are matching VolumeClaimTemplate specification.
If not, a call is made to update the PVC object with the size specified in the VolumeClaimTemplate.

### RBAC changes.
Now that StatefulSet controller needs to update the PVC, `update` will be added to the `statefulset-controller` role in the buildControllerRoles.

### Test Plan

> 

Reviewer: Add a test plan?

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

> Currently kubernetes supports volume expansion as a beta feature since v1.16. However, 

Reviewer 2: ```suggestion
Kubernetes has supported volume expansion as a beta feature since v1.16. However,
```

this feature support is only limited to standalone pods and deployments. Expansion of 
volumes associated with a Statefulset is not supported. This enhancement proposes to add 
the ability to resize volumes associated with a StatefulSet via modifications to its specification.

Users will be able to recover from volume expansion failures using the same mechanism 
as listed in [KEP 1790](../1790-recover-resize-failure/README.md).

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

> Stable persistent volumes can be created by specifying `volumeClaimTemplates` in `StatefulSet`
> [specification](https://kubernetes.io/docs/concepts/workloads/controllers/StatefulSet/).

Reviewer 2: ```suggestion
Stable persistent volumes can be created by specifying `.spec.volumeClaimTemplates`
for a [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/).
```

- Shrinking of volume is not allowed.
- Few environments require Pod restarts in order
to get access to the resized volume.
- Modifying the storage class of the volumeClaimTemplate
- Resizing a PVC associated with a statefulset directly by modifying the PVC spec will not affect the 
  Statefulset `volumeClaimTemplates`. 

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

Before we proceed to the proposal it is imperative to layout the various components involved
in creating and managing the persistent volumes associated with `StatefulSet`. 

- StatefulSet controller creates the PVCs from `volumeClaimTemplates` and associates them with the pods
spawned from it. Reconciliation logic runs for every change made to the StatefulSet
`Spec` and `Status` reflects the progress towards making the real world state equal to the
`Spec`.

> - The admission controller in the API server is responsible for restricting the fields

Reviewer 3: nit: it's not an admission controller - it's just the validation logic

Author: Changed it

which can be edited in a StatefulSet specification.
- The RBAC specification in the controller policy is responsible for adding the required 
RBAC permission for StatefulSet.
- The StatefulSet controller also supports revision control.

The following changes are proposed to achieve the goal:
1. Make the `volumeClaimTemplates` storage size editable by modification to the api server
   validation. Some validation checks will depend on the design decisions made in [KEP 1790](../1790-recover-resize-failure/README.md) to prevent user from shrinking the volume.
2. Add reconciliation of the associated PVC's size and the individual `volumeClaimTemplates` size into the 
   StatefulSet controller reconciliation.
3. Add PVC `update` privilege to the StatefulSet controller's clusterrole RBAC.

Once the above changes are made, modification to any of the `volumeClaimTemplates` storage
size by user will update the underlying PVC via the reconciliation performed by (2) above.

> The existing PVC resize workflow kicks in at this point and attempts to resize the volume.

Reviewer 4: We should specify how we mean to apply those changes to `volumeClaimTemplate`. Such as:

1. Are we going to perform changes to PVC one-by-one and wait for resize to be finished on PVCs? Is there going to be an status or something else added to statefulset to indicate that an expansion is happening?
2. What about volume types that only allow offline expansion? Is statefulset controller going to delete pods and expand the corresponding volume and bring them back?
3.

### Error indications to user

Whether a resize works or not would depend on many factors such as is `allowExpansion` set in the
storage class, does the underlying storage allow expansion etc. Such failures will be reported in
the corresponding PVC events. StatefulSet would only report any failures in the update operation

> performed to the PVC object.

Reviewer 2: If a resize error happens, does the control plane fix that? If not, what steps can someone with write access to a StatefulSet take in order to prompt a retry / ensure reconciliation of desired and actual state?

Reviewer 2: Answers to that might belong in the Caveats section.

Reviewer 4: Yes I think we need to clarify - what happens if user edits `volumeClaimTemplate` and increases PVC size but underlying volume plugin does not support expansion. Is the change blocked via admission or change is accepted but never applied?

### Revision control

StatefulSet supports [revision control](https://kubernetes.io/blog/2017/09/kubernetes-statefulsets-daemonsets/).
The storage size change will not be added to the fields supporting revision control. Any change in
revision - going forward to a new version or going back to an older version would not make any

> changes to the persistent volume size. Any changes requiring resize of the volume will need editing

Reviewer 3: Can you describe why you don't want to add that to revision control?

[My assumption is that adding it would effectively block rollback, as it would mean shrinking PVC. If that's the case, please add to the KEP. If not, please add your thinking behind that.]

Reviewer 4: What happens if user requests larger volume size and may be changes version of image in same edit? Are these operations going to be serialized one-after-other or applied at once for each replica of statefulset?

Statefulset controller revision code. This will facilitate cases where statefulset has to roll back to a previous 
version with smaller requested storage size without the need to shrink the volume. Detailed documentation will be 
added to mitigate any user expectation mismatch.

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### API server validation relaxation
Modifications would be made to `ValidateStatefulSetUpdate` in order to allow changes in the
storage size of the volume in the `VolumeClaimTemplates`. The new requested size will be validated to 
prevent a user from shrinking the volume.

### StatefulSet controller changes
The `sync` function in the `StatefulSetController` runs the reconciliation to match the current
reality and the expected declaration in the `Spec`. An addendum will be made to validate each
PVC associated with each pod and ensure that their sizes are matching VolumeClaimTemplate specification.

> If not, a call is made to update the PVC object with the size specified in the VolumeClaimTemplate.

Reviewer 5: This should probably be,
If the pvc's size is less then the size specified in the statefulset.spec, it will be updated.

If the user manually updated the pvc to a bigger value, this behavior will prevent an attempt to srhink.

Author: Updated it. Thank you!

* *GA* in 1.26  

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

### Version Skew Strategy

<!--
If applicable, how will the component handle version skew with other
components? What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- Does this enhancement involve coordinating behavior in the control plane and
  in the kubelet? How does an n-2 kubelet without this feature available behave
  when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

> ## Production Readiness Review Questionnaire

Reviewer 3: Feel free to add me as PRR approver for this KEP
